From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Fri, 29 Aug 2025 03:13:29 +0200
Subject: kms-connector: Only trigger a connector update on disconnect timeout

If a disconnection timeout happens we can just trigger a disconnect
update in thread, without bothering the other devices or connectors

Origin: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4610
Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/1607
Bug-Ubuntu: https://launchpad.net/bugs/1971434
---
 src/backends/native/meta-kms-connector.c | 31 ++++++++++++++++++++++++++++++-
 1 file changed, 30 insertions(+), 1 deletion(-)

diff --git a/src/backends/native/meta-kms-connector.c b/src/backends/native/meta-kms-connector.c
index ce76882..dc6203e 100644
--- a/src/backends/native/meta-kms-connector.c
+++ b/src/backends/native/meta-kms-connector.c
@@ -1041,6 +1041,35 @@ meta_kms_connector_update_state_changes (MetaKmsConnector       *connector,
     current_state->privacy_screen_state = new_state->privacy_screen_state;
 }
 
+static gpointer
+update_connector_in_impl (MetaThreadImpl  *thread_impl,
+                          gpointer         user_data,
+                          GError         **error)
+{
+  MetaKmsConnector *connector = user_data;
+  MetaKmsResourceChanges changes;
+
+  changes = meta_kms_impl_device_update_states (connector->impl_device,
+                                                0, connector->id);
+
+  return GUINT_TO_POINTER (changes);
+}
+
+static MetaKmsResourceChanges
+update_connector (MetaKmsConnector *connector)
+{
+  MetaKmsImpl *impl = meta_kms_impl_device_get_impl (connector->impl_device);
+  MetaKms *kms = meta_kms_impl_get_kms (impl);
+  g_autoptr (MetaKmsConnector) reffed_connector = g_object_ref (connector);
+  gpointer ret;
+
+  meta_assert_not_in_kms_impl (kms);
+
+  ret = meta_kms_run_impl_task_sync (kms, update_connector_in_impl,
+                                     reffed_connector, NULL);
+  return GPOINTER_TO_UINT (ret);
+}
+
 static gboolean
 connection_change_timeout (gpointer user_data)
 {
@@ -1065,7 +1094,7 @@ connection_change_timeout (gpointer user_data)
                   dispatch_time - ready_time);
     }
 
-  changes = meta_kms_update_states_sync (kms);
+  changes = update_connector (connector);
 
   meta_topic (META_DEBUG_KMS, "%s, changes=0x%x", G_STRFUNC, changes);
 
